packages/host/app/services/command-service.ts

```diff
@@ -142,14 +170,126 @@ export default class CommandService extends Service {
+
+        // Auto-execute if LLM mode is 'act' AND the command came after the LLM mode was set to 'act',
+        // or if requiresApproval is false
+        let shouldAutoExecute = false;
+
+        if (
+          messageCommand.requiresApproval === false ||
+          activeModeAtMessageTime === 'act'
+        ) {
+          shouldAutoExecute = true;
+        }
+
+        if (shouldAutoExecute) {
           this.run.perform(messageCommand);
         }
       }
     }
     finishedProcessingCommands!();
   }
 
+  private async drainCodePatchProcessingQueue() {
+    await this.flushCodePatchProcessingQueue;
+
+    let finishedProcessingCodePatches: () => void;
+    this.flushCodePatchProcessingQueue = new Promise(
+      (res) => (finishedProcessingCodePatches = res),
+    );
+
+    let codePatchSpecs = [...this.codePatchProcessingEventQueue];
+    this.codePatchProcessingEventQueue = [];
+
+    while (codePatchSpecs.length > 0) {
+      let [roomId, eventId] = codePatchSpecs.shift()!.split('|');
+
+      let roomResource = this.matrixService.roomResources.get(roomId!);
+      if (!roomResource) {
+        throw new Error(
+          `Room resource not found for room id ${roomId}, this should not happen`,
+        );
+      }
+      let timeout = Date.now() + 60_000; // reset the timer to avoid a long wait if the room resource is processing
+      let currentRoomProcessingTimestamp = roomResource.processingLastStartedAt;
+      while (
+        roomResource.isProcessing &&
+        currentRoomProcessingTimestamp ===
+          roomResource.processingLastStartedAt &&
+        Date.now() < timeout
+      ) {
+        // wait for the room resource to finish processing
+        await delay(100);
+      }
+      if (
+        roomResource.isProcessing &&
+        currentRoomProcessingTimestamp === roomResource.processingLastStartedAt
+      ) {
+        // room seems to be stuck processing, so we will log and skip this event
+        console.error(
+          `Room resource for room ${roomId} seems to be stuck processing, skipping code patch event ${eventId}`,
+        );
+        continue;
+      }
+      let message = roomResource.messages.find((m) => m.eventId === eventId);
+      if (!message) {
+        continue;
+      }
+      if (message.agentId !== this.matrixService.agentId) {
+        // This code patch was sent by another agent, so we will not auto-execute it
+        continue;
+      }
+
+      // Get the LLM mode that was active when this message was created
+      let messageTimestamp = message.created.getTime();
+      let activeModeAtMessageTime =
+        roomResource.getActiveLLMModeAtTimestamp(messageTimestamp);
+      console.log('activeModeAtMessageTime', activeModeAtMessageTime);
+      // Only auto-apply if in 'act' mode
+      if (activeModeAtMessageTime !== 'act') {
+        continue;
+      }
+
+      // Auto-apply all ready code patches from this message
```

Comment:
This must be redundant code for what is run when the user clicks Accept All. Can we consolidate?